File: src/p0318/java/solution.java
class Solution {
    public int maxProduct(String[] words) {
        int n = words.length;
        int[] comps = new int[n];
        int[] sizes = new int[n];
        int result = 0;
        for (int i = 0; i < n; i++) {
            int compi = 0, sizei = words[i].length();
            for (int k = 0; k < sizei; k++) {
                compi |= 1 << (words[i].charAt(k) - 'a');
            }
            comps[i] = compi;
            sizes[i] = sizei;
            for (int j = 0; j < i; j++) {
                if ((compi & comps[j]) == 0) {
                    result = Math.max(result, sizei * sizes[j]);
                }
            }
        }
        return result;
    }
}

File: src/p0318/cpp/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;


class Solution {
public:
    int maxProduct(vector<string> &words) {
        auto n = words.size();
        vector<int> comps(n), sizes(n);
        int result = 0;
        for (int i = 0; i < n; i++) {
            int compi = 0, sizei = int(words[i].size());
            for (auto c : words[i]) {
                compi |= (1 << (c - 'a'));
            }
            comps[i] = compi;
            sizes[i] = sizei;
            for (int j = 0; j < i; j++) {
                if (!(compi & comps[j])) {
                    result = max(result, sizei * sizes[j]);
                }
            }
        }
        return result;
    }
};

